sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -836,17 +836,30 @@ class Analyzer(
           Generate(newG.asInstanceOf[Generator], join, outer, qualifier, output, child)
         }
 
+      // If grouping keys have unresolved expressions, we need to replace them with resolved one
+      // in SELECT clauses.
+      case agg @ Aggregate(groups, aggs, child)
+          if conf.groupByAliasesEnabled && child.resolved && aggs.forall(_.resolved) &&
+            groups.exists(!_.resolved) =>
+        agg.copy(groupingExpressions = groups.map {
+            case u: UnresolvedAttribute =>
+              aggs.find(ne => resolver(ne.name, u.name)).map {
+                case Alias(e, _) => e
+                case e => e
+              }.getOrElse(u)
+            case e => e
+          })
+
       // Skips plan which contains deserializer expressions, as they should be resolved by another
       // rule: ResolveDeserializer.
       case plan if containsDeserializer(plan.expressions) => plan
 
       case q: LogicalPlan =>
         logTrace(s"Attempting to resolve ${q.simpleString}")
-        q transformExpressionsUp  {
+        q.transformExpressionsUp  {
           case u @ UnresolvedAttribute(nameParts) =>
-            // Leave unchanged if resolution fails.  Hopefully will be resolved next round.
-            val result =
-              withPosition(u) { q.resolveChildren(nameParts, resolver).getOrElse(u) }
+            // Leave unchanged if resolution fails. Hopefully will be resolved next round.
+            val result = withPosition(u) { q.resolveChildren(nameParts, resolver).getOrElse(u) }
```

Comment:
Ideally, we should first reach here to resolve them using the children, and then resolve the unresolved expression in grouping expression using the alias defined in aggregation expression. 

Currently, the order is wrong, but the guard `aggs.forall(_.resolved)` and `groups.exists(!_.resolved)` save us. I think we might need to separate it and add a new analyzer rule? 

cc @cloud-fan what is your opinion?